### ISSUE TITLE
OSC hides on cycle pause when showonpause=yes

### ISSUE THREAD

**Issue by Keith94** [Samillion/ModernZ#230]
### mpv Information

```bash
mpv 43d8966 Copyright © 2000-2024 mpv/MPlayer/mplayer2 projects
 built on Nov 11 2024 22:24:18
libplacebo version: v7.349.0
FFmpeg version: b862792
FFmpeg library versions:
   libavcodec      61.19.100
   libavdevice     61.3.100
   libavfilter     10.4.100
   libavformat     61.7.100
   libavutil       59.39.100
   libswresample   5.3.100
   libswscale      8.3.100
```

### Reproduction Steps

1. Open video file
2. Hover the osc to show it
3. Pause video and unpause it

### Expected Behavior

OSC respects hide timeout value set by script/user, same as stock osc behavior

### Actual Behavior

OSC hides immediately when resuming playback

### Log File

[output.txt](https://github.com/user-attachments/files/17788262/output.txt)

### I carefully read all instructions and confirm that I did the following:

- [X] I tested with the latest mpv version to validate that the issue is not already fixed.
- [X] I provided all required information.

**Comment by Keith94** [comment]
That is no worries.

When I tested it against stock OSC it didn't happen there. I thought it was a bug since hidetimeout is respected everywhere else.

**Comment by Samillion** [comment]
The logic is simple: on un-pause => reset timeout so it doesn't hide immediately, but nope, that doesn't work. Something is forcing the OSC to hide on un-pause, and I can't see it.

Auto hide logic properly re-arms:
https://github.com/Samillion/ModernZ/blob/322f20e078a848860bcc971c63deecf07ff44ed0/modernz.lua#L3148-L3168

Observing pause property doesn't force hide:
https://github.com/Samillion/ModernZ/blob/322f20e078a848860bcc971c63deecf07ff44ed0/modernz.lua#L3501-L3522

Sigh. I thought this would be a simple fix. I really think I fried my brain.

**Comment by Keith94** [comment]
That's unfortunate :( maybe I'll have a look at mpv codebase in case anything stands out

**Comment by Samillion** [comment]
![a9yqPXD_700b](https://github.com/user-attachments/assets/0d01bc7e-1e6c-4abe-a94b-44f6868362ae)

😆

**Comment by Keith94** [comment]
I made the following change to this function, and the behavior seems fixed, but I really want to know if this breaks anything.

```diff
local function enable_osc(enable)
    state.enabled = enable
    if enable then
-        do_enable_keybindings()
-    else
-        hide_osc() -- acts immediately when state.enabled == false
-        if state.showhide_enabled then
-            mp.disable_key_bindings("showhide")
-            mp.disable_key_bindings("showhide_wc")
-        end
-        state.showhide_enabled = false
+        show_osc()
    end
end
```

**Comment by Samillion** [comment]
Oooo, nice find. Removing all of them would definitely break many things (ie: clicking on stuff won't work).

Could you try just removing `hide_osc()` not sure why it's there in the first place when the function is about enabling the osc.

I'm not home at the moment, so I'm stuck relying on my lame logic lol.

**Comment by Samillion** [comment]
I'll ask on irc about why hide_osc() is added there as well, just to be sure.

**Comment by Keith94** [comment]
>clicking on stuff won't work

Hmm..clicks are working fine for me. I tested every button.

>Could you try just removing hide_osc()

If I remove that line, the osc still hides.

Let me know what you find out later on pls.

**Comment by Samillion** [comment]
Remove this as well, please: `state.showhide_enabled = false`

**Comment by Samillion** [comment]
Leaving this as a note: guidocella kindly pointed out the related commit that added this:
https://github.com/mpv-player/mpv/commit/3ab6af4f59dd21e3c9c514a25be380f883d608fe

**Comment by Keith94** [comment]
```lua
local function enable_osc(enable)
    state.enabled = enable
    if enable then
        do_enable_keybindings()
    else
        if state.showhide_enabled then
            mp.disable_key_bindings("showhide")
            mp.disable_key_bindings("showhide_wc")
        end
    end
end
```

Not working :(

**Comment by Samillion** [comment]
What the hell, why are element keybinds interfering with this or cause an immediate hide?

Either way, great find. I'll be back home by tomorrow, meanwhile, please test your initial adjustment to see if anything is out of the ordinary.

**Comment by Keith94** [comment]
>https://github.com/mpv-player/mpv/commit/3ab6af4f59dd21e3c9c514a25be380f883d608fe
it was displaying the osc briefly and  then autohide right away

Yeah, with my changes it causes this issue ^ (osc pops up on file loaded)

**Comment by Samillion** [comment]
> Edit: I simply added `show_osc()` instead of removing all that stuff to get the same effect

Awesome. I'll test with you as well when I'm back home. Well done, man. If all goes well, this should definitely be your PR.

**Comment by Keith94** [comment]
This seems to be related to the option `showonpause`. When having set `showonpause=no`, the bug is now fixed..

So it has something to do with that function where `showonpause` and `visibility_mode` are mentioned.

**Comment by Samillion** [comment]
The problem specifically is the call for `show_osc()` which triggers https://github.com/Samillion/ModernZ/issues/230#issuecomment-2567140739

As in: `state.enabled == false` => `hide_osc()`
https://github.com/Samillion/ModernZ/blob/d21d00f09325b931d596fca720b9dad9d6b56d5f/modernz.lua#L2991-L3003

**Comment by Keith94** [comment]
Ahh cool. so you know how to fix it or have some idea?

**Comment by Samillion** [comment]
The cause, yeah. Still trying to think of a way that wouldn't adjust the core code, Basically, if I modify, I want it to be in the `showonpause` part, not the `show_osc()`, `hide_osc()` or `enable_osc()` functions

**Comment by Samillion** [comment]
**Test script**: [show_on_pause/modernz.lua](https://github.com/Samillion/ModernZ/blob/dev_showonpause_fix_timeout/modernz.lua)
**Commit**: https://github.com/Samillion/ModernZ/commit/979c3c6e4e952af416ab67a14d65eda4ce2d8957

**Changes**:
- Respect hide timeout on unpause with `showonpause` enabled

**Comment by Samillion** [comment]
**Script**: [showonpause/modernz.lua](https://github.com/Samillion/ModernZ/blob/dev_pausecycle_showonpause/modernz.lua)
**Commit**: https://github.com/Samillion/ModernZ/commit/0ad50aa5053d991f932bd68ef4f64b5e8c94196f

**Changes**:
- Respect hide timeout on pause cycle when `showonpause=yes`

**Comment by Keith94** [comment]
How the heck did you find that!

Anyways, that definitely fixed it. Good to merge. 😸

**Comment by Samillion** [comment]
It was right in front of me all along, I even literally mentioned the fix but my stupid brain decided to vanish lol.

By far, I hate this bug the most because it shows how my brain just pretends to exist.

**Comment by Keith94** [comment]
Nah. I mean, if I can barely read 1% of the code, then I think you're totally fine lol.
